src/main/java/com/datatorrent/netlet/WriteOnlyClient.java

```diff
@@ -0,0 +1,239 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datatorrent.netlet;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.SocketChannel;
+import java.util.AbstractQueue;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.jctools.queues.SpscArrayQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datatorrent.netlet.util.Slice;
+
+public class WriteOnlyClient extends AbstractClientListener
+{
+  private static final Logger logger = LoggerFactory.getLogger(WriteOnlyClient.class);
+
+  private static class ForwardingQueue extends AbstractQueue<Slice>
+  {
+    private final Queue<Slice> queue;
+
+    private ForwardingQueue(Queue<Slice> queue)
+    {
+      this.queue = queue;
+    }
+
+    @Override
+    public Iterator<Slice> iterator()
+    {
+      return queue.iterator();
+    }
+
+    @Override
+    public int size()
+    {
+      return queue.size();
+    }
+
+    @Override
+    public boolean offer(Slice slice)
+    {
+      return queue.offer(slice);
+    }
+
+    @Override
+    public Slice poll()
+    {
+      return queue.poll();
+    }
+
+    @Override
+    public Slice peek()
+    {
+      return queue.peek();
+    }
+  }
+
+  protected final ByteBuffer writeBuffer;
+  protected Queue<Slice> sendQueue;
+  protected final SpscArrayQueue<Slice> freeQueue;
+  protected boolean isWriteEnabled = true;
+  protected Lock lock = new ReentrantLock();
+
+  public WriteOnlyClient()
+  {
+    this(64 * 1024, 1024);
+  }
+
+  public WriteOnlyClient(final int writeBufferCapacity, final int sendQueueCapacity)
+  {
+    this(ByteBuffer.allocateDirect(writeBufferCapacity), sendQueueCapacity);
+  }
+
+  public WriteOnlyClient(final ByteBuffer writeBuffer, final int sendQueueCapacity)
+  {
+    this.writeBuffer = writeBuffer;
+    sendQueue = new SpscArrayQueue<Slice>(sendQueueCapacity);
+    freeQueue = new SpscArrayQueue<Slice>(sendQueueCapacity);
+  }
+
+  @Override
+  public void connected()
+  {
+    super.connected();
+    shutdownIO(true);
+    suspendReadIfResumed();
+    try {
+      lock.lock();
+      isWriteEnabled = !isWriteSuspended();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void unregistered(SelectionKey key)
+  {
+    try {
+      lock.lock();
+      if (sendQueue.peek() == null && !isWriteEnabled) {
+        super.unregistered(key);
+      } else {
+        sendQueue = new ForwardingQueue(sendQueue)
```

Comment:
Assuming this is so that the client can write pending data without accepting a new data after a disconnect. In DefaultEventLoop, on disconnect, the listener is being replaced with a DisconnectingListener which gives client a chance to write out remaining data. I don't see where the key is getting cancelled and channel closed?